mmdet/models/dense_heads/maskformer_head.py

```diff
@@ -0,0 +1,709 @@
+            label_weights_list, dim=0)  # shape [bs, bq]
+        # ! shape [n_gts, h, w]
+        mask_targets = torch.cat(mask_targets_list, dim=0)
+        mask_weights = torch.stack(
+            mask_weights_list, dim=0)  # ! shape [bs, nq]
+
+        # classfication loss
+        cls_scores = cls_scores.flatten(0, 1)  # shape [bs * nq, ]
+        labels = labels.flatten(0, 1)  # shape [bs * nq, ]
+        label_weights = label_weights.flatten(0, 1)  # shape [bs* nq, ]
+
+        class_weight = torch.ones(self.num_classes + 1, device=labels.device)
+        class_weight[-1] = self.bg_cls_weight
+        loss_cls = self.loss_cls(
+            cls_scores,
+            labels,
+            label_weights,
+            avg_factor=class_weight[labels].sum())
+
+        num_total_masks = reduce_mean(cls_scores.new_tensor([num_total_pos]))
+        num_total_masks = max(num_total_masks, 1)
+
+        # extract positive ones
+        mask_preds = mask_preds[mask_weights > 0]
+        target_shape = mask_targets.shape[-2:]
+
+        if mask_targets.shape[0] == 0:
+            # zero match
+            loss_dice = mask_preds.sum()
+            loss_mask = mask_preds.sum()
+            return loss_cls, loss_mask, loss_dice
+
+        # upsample to shape of target
+        mask_preds = F.interpolate(
+            mask_preds.unsqueeze(1),
+            target_shape,
+            mode='bilinear',
+            align_corners=False).squeeze(1)  # [n_gts, h, w]
+
+        # dice loss
+        loss_dice = self.loss_dice(
+            mask_preds, mask_targets, avg_factor=num_total_masks)
+
+        # mask loss
+        # ! FocalLoss support input of shape [n, num_class]
+        h, w = mask_preds.shape[-2:]
+        # [n_gts, h, w] -> [n_gts * h * w, 1]
+        mask_preds = mask_preds.reshape(-1, 1)
+        # [n_gts, h, w] -> [n_gts * h * w]
+        mask_targets = mask_targets.reshape(-1)
+        # ! 1 - mask_targets
+        loss_mask = self.loss_mask(
+            mask_preds, 1 - mask_targets, avg_factor=num_total_masks * h * w)
+
+        return loss_cls, loss_mask, loss_dice
+
+    def forward(self, feats, img_metas):
+        """Forward function.
+
+        Args:
+            feats (list[Tensor]): Features from the upstream network, each
+                is a 4D-tensor.
+            img_metas (list[dict]): List of image information.
+
+        Returns:
+            all_cls_scores (Tensor): Classification scores for each
+                scale level. Each is a 4D-tensor with shape
+                    [nb_dec, bs, num_query, cls_out_channels].
+                 Note `cls_out_channels` should includes background.
+            all_mask_preds (Tensor): Mask scores for each decoder
+                layer. Each with shape [n_dec, bs, num_query, h, w].
+        """
+        bs = len(img_metas)
+        input_img_h, input_img_w = img_metas[0]['batch_input_shape']
+        padding_mask = feats[-1].new_zeros((bs, input_img_h, input_img_w),
```

Comment:
here we should use new_ones and assign the unpadded area as 0, please check the docstring of multihead attention about the meaning and value of padding mask